core/ev-admin.php

```diff
@@ -0,0 +1,1119 @@
+
+    if( update_option( 'sp_external_videos_options', $options ) ) $message = "Settings saved.";
+    ob_clean();
+
+    wp_send_json( $message );
+
+  }
+
+  /*
+  *  update_videos_handler
+  *
+  *  Used by settings page
+  *  AJAX handler for "Update videos from channels" form
+  *
+  *  @type  function
+  *  @date  31/10/16
+  *  @since  1.0
+  *
+  *  @param
+  *  @return
+  */
+
+  function update_videos_handler() {
+
+    // Handle the ajax request
+    check_ajax_referer( 'ev_settings' );
+
+    // Get options once now for the helper functions
+    $options = SP_External_Videos::get_options();
+    $AUTHORS = $options['authors'];
+    $HOSTS = $options['hosts'];
+    $delete = isset( $options['delete'] ) ? $options['delete'] : null;
+
+    $new_messages = $trash_messages = '';
+
+    // figure out whether we're updating a single author, or all
+    // and populate the $update_authors array
+    if( isset( $_POST['author_id'] ) && isset( $_POST['host_id'] ) ) {
+      // it's single
+      $eligibles = array();
+
+      // get the relevant local author. not so simple in multidimensional!
+      foreach( $AUTHORS as $authorkey => $authordata ){
+        if( in_array( $_POST['host_id'], $authordata ) ) {
+          $eligibles[] = $authorkey;
+        }
+      }
+
+      if( $eligibles ) {
+        $thisindex = array();
+
+        foreach( $eligibles as $eligible => $originalkey ){
+          if( in_array( $_POST['author_id'], $AUTHORS[$originalkey] ) ) {
+            $thisindex[] = $originalkey;
+          }
+        }
+        $thisindex = array_shift( $thisindex );
+        $update_authors = array( $AUTHORS[$thisindex] );
+        $single = true;
+      }
+
+    } else { // it's update all
+      $update_authors = $AUTHORS;
+      $single = null;
+    }
+
+    // post_new_videos() gets everything new and returns messages about it
+    $new_messages = $this->post_new_videos( $update_authors, $HOSTS, $single );
+
+    // trash_deleted_videos() checks for videos deleted on host and returns messages about it
+    if( $delete && !$single ) $trash_messages = $this->trash_deleted_videos( $update_authors, $HOSTS );
+
+    $messages = $new_messages . $trash_messages;
+    wp_send_json( $messages );
+
+  }
+
+  /*
+  *  post_new_videos
+  *
+  *  Used by update_videos_handler() and daily_function()
+  *  Saves any new videos from host channels to the database.
+  *  Returns messages about number of video posts added.
+  *  Works for single-author and update-all via $single param
+  *
+  *  @type  function
+  *  @date  31/10/16
+  *  @since  1.0
+  *
+  *  @param   $authors, $HOSTS, $single
+  *  @return  html $messages
+  */
+
+  function post_new_videos( $authors, $HOSTS, $single ) {
+
+    $new_videos = $this->fetch_new_videos( $authors, $HOSTS );
+
+    $messages = $add_messages = $no_messages = $zero_message = '';
+
+    if ( !$new_videos ) {
+      $zero_message = "No videos found.";
```

Comment:
Hmm, you've circumvented the problem with error messages from the hosts by simply saying "No videos found" rather than displaying the error reported, e.g. on Vimeo you could get this as a reply: 

``Array
(
    [error] => You must provide a valid authenticated access token.
)``

but we now simply say "No videos found". I think we can handle this better.

For exampel, I had suggested creating exceptions in ev-vimeo.php when an error is returned and handling the exception in ev-admin.php.

If I'm not explaining this well enough, I can fix that up after merging, but I'd be happy for you to give it a go.